cli/azd/internal/telemetry/tracer.go

```diff
@@ -0,0 +1,138 @@
+	"go.opentelemetry.io/otel/trace"
+)
+
+// It is often valuable to extend functionality of 3rd-party library types.
+// Therefore, we provide extension points for OpenTelemetry tracers here.
+
+// Tracer is the creator of Spans.
+//
+// This is simply trace.Tracer except returning our Span instead of trace.Span.
+type Tracer interface {
+	Start(ctx context.Context, spanName string, opts ...trace.SpanStartOption) (context.Context, Span)
+}
+
+// Wrapper around trace.Tracer.
+type wrapperTracer struct {
+	tracer trace.Tracer
+}
+
+func (w *wrapperTracer) Start(ctx context.Context, spanName string, opts ...trace.SpanStartOption) (context.Context, Span) {
+	ctx, span := w.tracer.Start(ctx, spanName, opts...)
+	return ctx, &wrapperSpan{span}
+}
+
+// redefinedSpan is a slightly modified version of trace.Span.
+//
+// The only changes made is to remove functionality around emitting events.
+// Events are nested telemetry events that can be fired from a Span.
+// We currently do not support this yet (no use case), but this is likely to be changed in the future.
+//
+// Current modifications to trace.Span:
+//   - Removed AddEvent
+//   - Removed RecordError. This creates an error type event
+type redefinedSpan interface {
+	// End completes the Span. The Span is considered complete and ready to be
+	// delivered through the rest of the telemetry pipeline after this method
+	// is called. Therefore, updates to the Span are not allowed after this
+	// method has been called.
+	End(options ...trace.SpanEndOption)
+
+	// IsRecording returns the recording state of the Span. It will return
+	// true if the Span is active and events can be recorded.
+	IsRecording() bool
+
+	// SpanContext returns the SpanContext of the Span. The returned SpanContext
+	// is usable even after the End method has been called for the Span.
+	SpanContext() trace.SpanContext
+
+	// SetStatus sets the status of the Span in the form of a code and a
+	// description, overriding previous values set. The description is only
+	// included in a status when the code is for an error.
+	SetStatus(code codes.Code, description string)
+
+	// SetName sets the Span name.
+	SetName(name string)
+
+	// SetAttributes sets kv as attributes of the Span. If a key from kv
+	// already exists for an attribute of the Span it will be overwritten with
+	// the value contained in kv.
+	SetAttributes(kv ...attribute.KeyValue)
+
+	// TracerProvider returns a TracerProvider that can be used to generate
+	// additional Spans on the same telemetry pipeline as the current Span.
+	TracerProvider() trace.TracerProvider
+}
+
+// Span is the individual component of a trace. It represents a single named
+// and timed operation of a workflow that is traced. A Tracer is used to
+// create a Span and it is then up to the operation the Span represents to
+// properly end the Span when the operation itself ends.
+type Span interface {
+	redefinedSpan
+}
+
+type wrapperSpan struct {
+	span trace.Span
+}
```

Comment:
What is the use case for the `wrapperSpan` vs the standard `Span` ?

Reply:
OK - so this is just wrapping the otel `Span` ?

Reply:
Yep, for any extended functionality